src/sync/generic_network_transport.hpp

```diff
@@ -0,0 +1,324 @@
+        else if (code == "FunctionDuplicateName")
+            return service_error_code::function_duplicate_name;
+        else if (code == "FunctionSyntaxError")
+            return service_error_code::function_syntax_error;
+        else if (code == "FunctionInvalid")
+            return service_error_code::function_invalid;
+        else if (code == "IncomingWebhookNotFound")
+            return service_error_code::incoming_webhook_not_found;
+        else if (code == "IncomingWebhookAlreadyExists")
+            return service_error_code::incoming_webhook_already_exists;
+        else if (code == "IncomingWebhookDuplicateName")
+            return service_error_code::incoming_webhook_duplicate_name;
+        else if (code == "RuleNotFound")
+            return service_error_code::rule_not_found;
+        else if (code == "APIKeyNotFound")
+            return service_error_code::api_key_not_found;
+        else if (code == "RuleAlreadyExists")
+            return service_error_code::rule_already_exists;
+        else if (code == "AuthProviderDuplicateName")
+            return service_error_code::auth_provider_duplicate_name;
+        else if (code == "RestrictedHost")
+            return service_error_code::restricted_host;
+        else if (code == "APIKeyAlreadyExists")
+            return service_error_code::api_key_already_exists;
+        else if (code == "IncomingWebhookAuthFailed")
+            return service_error_code::incoming_webhook_auth_failed;
+        else if (code == "ExecutionTimeLimitExceeded")
+            return service_error_code::execution_time_limit_exceeded;
+        else if (code == "NotCallable")
+            return service_error_code::not_callable;
+        else if (code == "UserAlreadyConfirmed")
+            return service_error_code::user_already_confirmed;
+        else if (code == "UserNotFound")
+            return service_error_code::user_not_found;
+        else if (code == "UserDisabled")
+            return service_error_code::user_disabled;
+        else
+            return service_error_code::unknown;
+    }
+
+    const service_error_code m_code;
+    const std::string m_raw_code;
+};
+
+} // namespace error
+
+
+enum class Method {
+    get, post, patch
+};
+
+/**
+ * An HTTP request that can be made to an arbitrary server.
+ */
+struct Request {
+public:
+    /**
+     * The HTTP method of this request.
+     */
+    Method method;
+
+    /**
+     * The URL to which this request will be made.
+     */
+    std::string url;
+
+    /**
+     * The number of seconds that the underlying transport should spend on an HTTP round trip before failing with an
+     * error.
+     */
+    int timeout_ms;
+
+    /**
+     * The HTTP headers of this request.
+     */
+    std::map<std::string, std::string> headers;
+
+    /**
+     * The body of the request.
+     */
+    std::string body;
+};
+
+/**
+ * The contents of an HTTP response.
+ */
+struct Response {
```

Comment:
This response also needs to be able to handle other error types than just HTTP error codes, e.g. a socket timeout exception or some other exception thrown from the binding. It is a bit unclear exactly how this should be modeled though, because most likely in Java you want to roundtrip the same exception.

We could probably encode the semantics using special status_codes not used by REST API's and encode the message in body, but it probably does make for a slightly magical API. 

Would it make sense to have a both a `http_status_code` and `binding_status_code` property. Bindings could then define their own semantics to binding_status_codes?

Reply:
I've modelled this after what we had previously, but we can add binding_status_code as well.

Reply:
We don't use the terminology "binding", but SDK. When it slips, it should at least not go into the code :-)

Reply:
Could just call it as a custom status code then